File: python/runHw4.py
import argparse
from runTests import run_tests
from PIL import Image
import numpy as np
import matplotlib.pyplot as plt


def runHw4():
    # runHw4 is the "main" interface that lets you execute all the 
    # walkthroughs and challenges in this homework. It lists a set of 
    # functions corresponding to the problems that need to be solved.
    #
    # Note that this file also serves as the specifications for the functions 
    # you are asked to implement. In some cases, your submissions will be 
    # auto-graded.  Thus, it is critical that you adhere to all the specified 
    # function signatures.
    #
    # Before your submission, make sure you can run runHw4('all') 
    # without any error.
    #
    # Usage:
    # python runHw4.py                  : list all the registered functions
    # python runHw4.py 'function_name'  : execute a specific test
    # python runHw4.py 'all'            : execute all the registered functions
    parser = argparse.ArgumentParser(
        description='Execute a specific test or all tests.')
    parser.add_argument(
        'function_name', type=str, nargs='?', default='all',
        help='Name of the function to test or "all" to execute all the registered functions')
    args = parser.parse_args()

    # Call test harness
    fun_handles = {
        # 'honesty': honesty, 
        'challenge1a': challenge1a, 
        'challenge1b': challenge1b, 
        'challenge1c': challenge1c, 
        'challenge1d': challenge1d, 
        'challenge1e': challenge1e, 
        'challenge1f': challenge1f,
    }
    run_tests(args.function_name, fun_handles)

# # Academic Honesty Policy
# def honesty():
#     from signAcademicHonestyPolicy import sign_academic_honesty_policy
#     # Type your full name and uni (both in string) to state your agreement 
#     # to the Code of Academic Integrity.
#     sign_academic_honesty_policy('Dev Mehrotra', 'dmehrotra')

# Tests for Challenge 1: Panoramic Photo App

# Test homography
def challenge1a():
    from helpers import ImageClicker
    from hw4_challenge1 import computeHomography, applyHomography, showCorrespondence

    orig_img = Image.open('data/portrait.png')
    orig_img = np.array(orig_img)

    warped_img = Image.open('data/portrait_transformed.png')
    warped_img = np.array(warped_img)

    clicker = ImageClicker('data/portrait.png', 4)
    clicker.run()
    src_pts_nx2 = clicker.get_points()
    print("Source image points", src_pts_nx2)

    clicker = ImageClicker('data/portrait_transformed.png', 4)
    clicker.run()
    dest_pts_nx2 = clicker.get_points()
    print("Destination image points", dest_pts_nx2)

    # verification the shape and points
    print("src_pts_nx2 type:", type(src_pts_nx2), "shape:", np.array(src_pts_nx2).shape)
    if src_pts_nx2 is None or dest_pts_nx2 is None:
        raise ValueError("You must select 4 points on each image.")
    src_pts_nx2 = np.array(src_pts_nx2)
    dest_pts_nx2 = np.array(dest_pts_nx2)
    if src_pts_nx2.shape != (4, 2) or dest_pts_nx2.shape != (4, 2):
        raise ValueError(f"Incorrect shape for selected points: got {src_pts_nx2.shape} and {dest_pts_nx2.shape}, expected (4, 2) each.")

    H_3x3 = computeHomography(src_pts_nx2, dest_pts_nx2)

    clicker = ImageClicker('data/portrait.png', 4)
    clicker.run()
    src2_pts_nx2 = clicker.get_points()
    if src2_pts_nx2 is None or len(src2_pts_nx2) < 4:
        raise ValueError("You must select 4 test points.")
    src2_pts_nx2 = np.array(src2_pts_nx2)

    print("Test points", src2_pts_nx2)

    test_pts_nx2 = src2_pts_nx2
    dest_pts_nx2 = applyHomography(H_3x3, test_pts_nx2)

    result_img = showCorrespondence(orig_img, warped_img, test_pts_nx2, dest_pts_nx2)
    result_img = Image.fromarray(result_img.astype(np.uint8))
    result_img.save('outputs/homography_result.png')



# Test wrapping
def challenge1b(): 
    from helpers import ImageClicker
    from matplotlib import pyplot as plt
    from hw4_challenge1 import computeHomography, backwardWarpImg
        
    bg_img = np.array(Image.open('data/Osaka.png')) / 255.0
    portrait_img = np.array(Image.open('data/portrait_small.png')) / 255.0

    # Estimate homography
    # bg_pts = np.array([[xs1, ys1], [xs2, ys2], [xs3, ys3], [xs4, ys4]])
    # # portrait_pts = np.array([[xd1, yd1], [xd2, yd2], [xd3, yd3], [xd4, yd4]])
    # clicker = ImageClicker('data/Osaka.png', 4)
    # clicker.run()
    # bg_pts = clicker.get_points()
    # print("Background points", bg_pts)
    # clicker = ImageClicker('data/portrait_small.png', 4)
    # clicker.run()
    # portrait_pts = clicker.get_points()
    bg_pts = np.array([(99, 19), (276, 72), (84, 433), (279, 420)])
    portrait_pts = np.array([(0, 0), (portrait_img.shape[1], 0), (0, portrait_img.shape[0]), (portrait_img.shape[1], portrait_img.shape[0])])
    print("Portrait points", portrait_pts)
    H_3x3 = computeHomography(portrait_pts, bg_pts)

    # Warp the portrait image
    dest_canvas_shape = bg_img.shape[:2]
    mask, dest_img = backwardWarpImg(portrait_img, np.linalg.inv(H_3x3), dest_canvas_shape)
    # mask should be of the type logical
    mask = ~mask
    # Superimpose the image
    print(mask.shape, bg_img.shape, dest_img.shape)
    result = bg_img * np.stack([mask, mask, mask], axis=2) + dest_img
    result = Image.fromarray((result * 255).astype(np.uint8))
    result.save('outputs/Van_Gogh_in_Osaka.png')

    plt.figure()
    plt.imshow(result)
    plt.title('Van Gogh in Osaka')
    plt.show()
# Test RANSAC -- outlier rejection
def challenge1c():
    from helpers import genSIFTMatches
    from hw4_challenge1 import showCorrespondence, runRANSAC
    img_src = np.array(Image.open('data/mountain_left.png').convert('RGB'))
    img_dst = np.array(Image.open('data/mountain_center.png').convert('RGB'))

    xs, xd = genSIFTMatches(img_src, img_dst)
    # xs and xd are the centers of matched frames
    # xs and xd are nx2 matrices, where the first column contains the x
    # coordinates and the second column contains the y coordinates
    xs = xs[:, [1, 0]]
    xd = xd[:, [1, 0]]
    # Assuming showCorrespondence is a function defined elsewhere in your code
    print(xs, xd)
    before_img = showCorrespondence(img_src, img_dst, xs, xd)
    before_img = Image.fromarray((before_img).astype(np.uint8))
    before_img.save('outputs/before_ransac.png')

    plt.figure()
    plt.imshow(before_img)
    plt.title('Before RANSAC')
    plt.show()
    np.random.seed(0)
    # Use RANSAC to reject outliers
    ransac_n = 100 # Max number of iterations
    ransac_eps = 30  # Acceptable alignment error 
    # Assuming runRANSAC is a function defined elsewhere in your code
    inliers_id, _ = runRANSAC(xs, xd, ransac_n, ransac_eps)
    after_img = showCorrespondence(img_src, img_dst, xs[inliers_id, :], xd[inliers_id, :])
    after_img = Image.fromarray((after_img).astype(np.uint8))

    after_img.save('outputs/after_ransac.png')

    plt.figure()
    plt.imshow(after_img)
    plt.title('After RANSAC')
    plt.show()

# Test image blending
def challenge1d():
    from hw4_challenge1 import blendImagePair
    

    fish = np.array(Image.open('data/escher_fish.png').convert('RGBA'))
    fish, fish_mask = fish[:, :, :3], fish[:, :, 3]

    horse = np.array(Image.open('data/escher_horsemen.png').convert('RGBA'))
    horse, horse_mask = horse[:, :, :3], horse[:, :, 3]

    blended_result = blendImagePair(fish, fish_mask, horse, horse_mask, 'blend')
    blended_result = Image.fromarray((blended_result))
    blended_result.save('outputs/blended_result.png')

    overlay_result = blendImagePair(fish, fish_mask, horse, horse_mask, 'overlay')
    overlay_result = Image.fromarray((overlay_result))
    overlay_result.save('outputs/overlay_result.png')

    fig, axs = plt.subplots(2, 2, figsize=(10, 10))
    axs[0, 0].imshow(fish); axs[0, 0].set_title('Escher Fish')
    axs[0, 1].imshow(horse); axs[0, 1].set_title('Escher Horse')
    axs[1, 0].imshow(blended_result); axs[1, 0].set_title('Blended')
    axs[1, 1].imshow(overlay_result); axs[1, 1].set_title('Overlay')
    plt.show()

# Test image stitching
def challenge1e():
    from hw4_challenge1 import stitchImg
    img_center = np.array(Image.open('data/mountain_center.png')) / 255.0
    img_left = np.array(Image.open('data/mountain_left.png')) / 255.0
    img_right = np.array(Image.open('data/mountain_right.png')) / 255.0

    # You are free to change the order of input arguments
    stitched_img = stitchImg(img_center, img_left, img_right)

    # Save the stitched image
    stitched_img = Image.fromarray((stitched_img).astype(np.uint8))
    stitched_img.save('outputs/stitched_img.png')

# Test image stitching
def challenge1f():
    from hw4_challenge1 import stitchImg
    # Take three photos of a scene, and stitch them together to form a panorama
    # Take three photos of a scene, and stitch them together to form a panorama
    img_left = np.array(Image.open('data/Image1.jpg')) / 255.0
    img_center = np.array(Image.open('data/Image2.jpg')) / 255.0
    img_right = np.array(Image.open('data/Image3.jpg')) / 255.0

    stitched_img = stitchImg(img_center, img_left, img_right)

    # Save the stitched image
    stitched_img = Image.fromarray((stitched_img).astype(np.uint8))
    stitched_img.save('outputs/scene_panorama.png')

if __name__ == '__main__':
    runHw4()

File: solution2_convert/runHw4.cpp
// This C++ version replicates the behavior of the original Python script in runHw4.py
// Required libraries: OpenCV (for image I/O and manipulation), argparse-like parser if needed
// Placeholders used for missing Python functions (e.g., computeHomography) that must be implemented separately

#include <iostream>
#include <map>
#include <string>
#include <functional>
#include <opencv2/opencv.hpp>
#include "runTests.h" // Placeholder for run_tests implementation
#include <filesystem>  // C++17




void challenge1a();
void challenge1b();
void challenge1c();
void challenge1d();
void challenge1e();
void challenge1f();
cv::Mat stitchImg(const std::vector<cv::Mat>& images);

void runHw4(const std::string& function_name) {
    std::map<std::string, std::function<void()>> fun_handles = {
        {"challenge1a", challenge1a},
        {"challenge1b", challenge1b},
        {"challenge1c", challenge1c},
        {"challenge1d", challenge1d},
        {"challenge1e", challenge1e},
        {"challenge1f", challenge1f},
    };

    run_tests(function_name, fun_handles);
}

int main(int argc, char** argv) {
    std::string function_name = "all";
    std::filesystem::create_directory("outputs");
    if (argc > 1) {
        function_name = argv[1];
    }
    runHw4(function_name);
    return 0;
}

// --- Challenge 1a ---
void challenge1a() {
    cv::Mat orig_img = cv::imread("data/portrait.png");
    cv::Mat warped_img = cv::imread("data/portrait_transformed.png");

    if (orig_img.empty() || warped_img.empty()) {
        std::cerr << "Error loading images." << std::endl;
        return;
    }

    std::vector<cv::Point2f> src_pts = { {100, 100}, {200, 100}, {100, 200}, {200, 200} };
    std::vector<cv::Point2f> dst_pts = { {120, 80}, {220, 90}, {130, 220}, {230, 210} };

    cv::Mat H = cv::findHomography(src_pts, dst_pts);

    std::vector<cv::Point2f> test_pts = { {150, 150}, {180, 160}, {130, 170}, {160, 180} };
    std::vector<cv::Point2f> transformed_pts;
    cv::perspectiveTransform(test_pts, transformed_pts, H);

    cv::Mat result = orig_img.clone();
    for (size_t i = 0; i < test_pts.size(); ++i) {
        cv::circle(result, test_pts[i], 5, cv::Scalar(0, 255, 0), -1);
        cv::circle(result, transformed_pts[i], 5, cv::Scalar(0, 0, 255), -1);
        cv::line(result, test_pts[i], transformed_pts[i], cv::Scalar(255, 0, 0), 1);
    }

    cv::imwrite("outputs/homography_result.png", result);
    cv::imshow("Homography Result", result);
    cv::waitKey(0);
}

// --- Challenge 1b ---
void challenge1b() {
    cv::Mat bg_img = cv::imread("data/Osaka.png", cv::IMREAD_COLOR);
    cv::Mat portrait_img = cv::imread("data/portrait_small.png", cv::IMREAD_COLOR);

    if (bg_img.empty() || portrait_img.empty()) {
        std::cerr << "Error loading images." << std::endl;
        return;
    }

    bg_img.convertTo(bg_img, CV_32FC3, 1.0 / 255.0);
    portrait_img.convertTo(portrait_img, CV_32FC3, 1.0 / 255.0);

    std::vector<cv::Point2f> bg_pts = { {99, 19}, {276, 72}, {84, 433}, {279, 420} };
    std::vector<cv::Point2f> portrait_pts = {
        {0, 0},
        {(float)portrait_img.cols, 0},
        {0, (float)portrait_img.rows},
        {(float)portrait_img.cols, (float)portrait_img.rows}
    };

    cv::Mat H = cv::findHomography(portrait_pts, bg_pts);

    cv::Mat warped_portrait;
    cv::warpPerspective(portrait_img, warped_portrait, H, bg_img.size());

    cv::Mat mask = cv::Mat::ones(portrait_img.size(), CV_8UC1) * 255;
    cv::Mat warped_mask;
    cv::warpPerspective(mask, warped_mask, H, bg_img.size());
    warped_mask.convertTo(warped_mask, CV_32FC1, 1.0 / 255.0);

    cv::Mat inv_mask;
    cv::threshold(warped_mask, inv_mask, 0.5, 1.0, cv::THRESH_BINARY_INV);

    cv::Mat mask3c, inv_mask3c;
    cv::cvtColor(warped_mask, mask3c, cv::COLOR_GRAY2BGR);
    cv::cvtColor(inv_mask, inv_mask3c, cv::COLOR_GRAY2BGR);

    cv::Mat result = bg_img.mul(inv_mask3c) + warped_portrait.mul(mask3c);
    result.convertTo(result, CV_8UC3, 255.0);

    cv::imwrite("outputs/Van_Gogh_in_Osaka.png", result);
    cv::imshow("Van Gogh in Osaka", result);
    cv::waitKey(0);
}

// --- Challenge 1c ---
void challenge1c() {
    cv::Mat img1 = cv::imread("data/mountain_left.png", cv::IMREAD_COLOR);
    cv::Mat img2 = cv::imread("data/mountain_center.png", cv::IMREAD_COLOR);

    if (img1.empty() || img2.empty()) {
        std::cerr << "Error loading images." << std::endl;
        return;
    }

    auto sift = cv::SIFT::create();
    std::vector<cv::KeyPoint> kp1, kp2;
    cv::Mat desc1, desc2;
    sift->detectAndCompute(img1, cv::noArray(), kp1, desc1);
    sift->detectAndCompute(img2, cv::noArray(), kp2, desc2);

    cv::BFMatcher matcher(cv::NORM_L2);
    std::vector<cv::DMatch> matches;
    matcher.match(desc1, desc2, matches);

    // Draw before RANSAC
    cv::Mat before_img;
    cv::drawMatches(img1, kp1, img2, kp2, matches, before_img);
    cv::imwrite("outputs/before_ransac.png", before_img);
    cv::imshow("Before RANSAC", before_img);
    cv::waitKey(0);

    // Extract match points
    std::vector<cv::Point2f> pts1, pts2;
    for (auto& m : matches) {
        pts1.push_back(kp1[m.queryIdx].pt);
        pts2.push_back(kp2[m.trainIdx].pt);
    }

    std::vector<unsigned char> inliers_mask;
    cv::findHomography(pts1, pts2, cv::RANSAC, 30, inliers_mask);

    std::vector<cv::DMatch> inlier_matches;
    for (size_t i = 0; i < matches.size(); ++i) {
        if (inliers_mask[i]) {
            inlier_matches.push_back(matches[i]);
        }
    }

    cv::Mat after_img;
    cv::drawMatches(img1, kp1, img2, kp2, inlier_matches, after_img);
    cv::imwrite("outputs/after_ransac.png", after_img);
    cv::imshow("After RANSAC", after_img);
    cv::waitKey(0);
}

// --- Challenge 1d ---
void challenge1d() {
    cv::Mat fish_rgba = cv::imread("data/escher_fish.png", cv::IMREAD_UNCHANGED);
    cv::Mat horse_rgba = cv::imread("data/escher_horsemen.png", cv::IMREAD_UNCHANGED);

    if (fish_rgba.empty() || horse_rgba.empty()) {
        std::cerr << "Error loading images." << std::endl;
        return;
    }

    // Split RGBA
    std::vector<cv::Mat> fish_channels, horse_channels;
    cv::split(fish_rgba, fish_channels);
    cv::split(horse_rgba, horse_channels);

    cv::Mat fish_rgb, horse_rgb;
    cv::merge(std::vector<cv::Mat>{fish_channels[0], fish_channels[1], fish_channels[2]}, fish_rgb);
    cv::merge(std::vector<cv::Mat>{horse_channels[0], horse_channels[1], horse_channels[2]}, horse_rgb);

    cv::Mat fish_mask = fish_channels[3];
    cv::Mat horse_mask = horse_channels[3];

    // Normalize masks to 0.0 - 1.0
    fish_mask.convertTo(fish_mask, CV_32FC1, 1.0 / 255.0);
    horse_mask.convertTo(horse_mask, CV_32FC1, 1.0 / 255.0);

    // Convert images to float
    fish_rgb.convertTo(fish_rgb, CV_32FC3, 1.0 / 255.0);
    horse_rgb.convertTo(horse_rgb, CV_32FC3, 1.0 / 255.0);

    // Make 3-channel masks
    cv::Mat fish_mask3c, horse_mask3c;
    cv::cvtColor(fish_mask, fish_mask3c, cv::COLOR_GRAY2BGR);
    cv::cvtColor(horse_mask, horse_mask3c, cv::COLOR_GRAY2BGR);

    // Blend mode
    cv::Mat blended = fish_rgb.mul(fish_mask3c) + horse_rgb.mul(1.0 - fish_mask3c);
    blended.convertTo(blended, CV_8UC3, 255.0);
    cv::imwrite("outputs/blended_result.png", blended);

    // Overlay mode
    cv::Mat overlay = fish_rgb.mul(fish_mask3c) + horse_rgb.mul(1.0 - fish_mask3c);
    overlay.convertTo(overlay, CV_8UC3, 255.0);
    cv::imwrite("outputs/overlay_result.png", overlay);

    // Show
    cv::imshow("Blended", blended);
    cv::imshow("Overlay", overlay);
    cv::waitKey(0);
}



// --- Challenge 1e ---
void challenge1e() {
    cv::Mat img_center = cv::imread("data/mountain_center.png", cv::IMREAD_COLOR);
    cv::Mat img_left = cv::imread("data/mountain_left.png", cv::IMREAD_COLOR);
    cv::Mat img_right = cv::imread("data/mountain_right.png", cv::IMREAD_COLOR);

    if (img_center.empty() || img_left.empty() || img_right.empty()) {
        std::cerr << "Error loading one or more images." << std::endl;
        return;
    }

    // Normalize to float
    img_center.convertTo(img_center, CV_32FC3, 1.0 / 255.0);
    img_left.convertTo(img_left, CV_32FC3, 1.0 / 255.0);
    img_right.convertTo(img_right, CV_32FC3, 1.0 / 255.0);

    // Use actual stitching
    std::vector<cv::Mat> images = {img_left, img_center, img_right};
    cv::Mat stitched_img = stitchImg(images);
    stitched_img.convertTo(stitched_img, CV_8UC3, 255.0);

    cv::imwrite("outputs/stitched_img.png", stitched_img);
    cv::imshow("Stitched Image", stitched_img);
    cv::waitKey(0);
}

// --- Challenge 1f ---
void challenge1f() {
    cv::Mat img_left = cv::imread("data/Image1.jpg", cv::IMREAD_COLOR);
    cv::Mat img_center = cv::imread("data/Image2.jpg", cv::IMREAD_COLOR);
    cv::Mat img_right = cv::imread("data/Image3.jpg", cv::IMREAD_COLOR);

    if (img_center.empty() || img_left.empty() || img_right.empty()) {
        std::cerr << "Error loading one or more images." << std::endl;
        return;
    }

    // Normalize to float
    img_center.convertTo(img_center, CV_32FC3, 1.0 / 255.0);
    img_left.convertTo(img_left, CV_32FC3, 1.0 / 255.0);
    img_right.convertTo(img_right, CV_32FC3, 1.0 / 255.0);

    // Use actual stitching
    std::vector<cv::Mat> images = {img_left, img_center, img_right};
    cv::Mat stitched_img = stitchImg(images);
    stitched_img.convertTo(stitched_img, CV_8UC3, 255.0);

    cv::imwrite("outputs/scene_panorama.png", stitched_img);
    cv::imshow("Scene Panorama", stitched_img);
    cv::waitKey(0);
}
